GVFS/GVFS.Common/Cleanup/GitCleanupScheduler.cs

```diff
@@ -0,0 +1,42 @@
+﻿using GVFS.Common.Git;
+using System;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupScheduler : IDisposable
+    {
+        private readonly TimeSpan prefetchPeriod = TimeSpan.FromMinutes(15);
+        private Timer prefetchStepTimer;
+        private GVFSContext context;
+        private GitObjects gitObjects;
+        private GitCleanupQueue queue;
+
+        public GitCleanupScheduler(GVFSContext context, GitObjects gitObjects, GitCleanupQueue queue)
+        {
+            this.context = context;
+            this.gitObjects = gitObjects;
+            this.queue = queue;
+
+            this.ScheduleRecurringCleanupSteps();
+        }
+
+        public void Dispose()
+        {
+            this.prefetchStepTimer?.Dispose();
+            this.prefetchStepTimer = null;
+        }
+
+        public void ScheduleRecurringCleanupSteps()
+        {
+            if (!this.context.Unattended && this.gitObjects.IsUsingCacheServer())
+            {
+                this.prefetchStepTimer = new Timer(
+                (state) => this.queue.Enqueue(new PrefetchStep(this.context, this.gitObjects, requireCacheLock: true)),
+                state: null,
+                dueTime: this.prefetchPeriod,
+                period: this.prefetchPeriod);

```

Comment:
If there are network issues or for some reason the step is taking longer than the 15 minutes could we end up with a bunch of steps enqueued?  Should it start timing after the step is completed?

Reply:
I think the steps should be responsible for checking if they "need to run" as the first thing inside of `Execute()`. This should be done under the lock. If we have five prefetch steps queued, then at least four of them will return quickly because the prefetch timestamp is too recent. Later steps will have similar ways to record when the last step of that type was run.

Note that it's not enough to limit the jobs in a single process, as multiple enlistments will be running steps at different times. If we have five enlistments trying to run background prefetch, the first to take the lock should win, and the rest give up (because of the lock or because of the recent timestamp).

Reply:
@wilbaker , we could simply check the queue for the same task before adding it right?  Or alternatively draining any dup cleanupEvents off the queue when we take one off.  

This isn't a bad idea if we think we're going to suffer from backlogs.  With todays design would we just have 2 cleanup jobs (potentially the same) running at the same time in that scenario.